src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2221,7 +2222,21 @@ RunDistributedExecution(DistributedExecution *execution)
 		/* always (re)build the wait event set the first time */
 		execution->rebuildWaitEventSet = true;
 
-		while (execution->unfinishedTaskCount > 0 && !cancellationReceived)
+		/*
+		 * Iterate until all the tasks are finished. Once all the tasks
+		 * are finished, ensure that that all the connection initializations
+		 * are also finished. Otherwise, those connections are terminated
+		 * abruptly before they are established (or failed). Instead, we let
+		 * the ConnectionStateMachine() to properly handle them.
+		 *
+		 * Note that we could have the connections that are not established
+		 * as a side effect of slow-start algorithm. At the time the algorithm
+		 * decides to establish new connections, the execution might have tasks
+		 * to finish. But, the execution might finish before the new connections
+		 * are established.
+		 */
+		while (execution->unfinishedTaskCount > 0 ||
+			   HasIncompleteConnectionEstablishment(execution))
```

Comment:
- Make some  perf tests
- What happens if connection timeout happens while `HasIncompleteConnectionEstablishment `

Reply:
@marcocitus, as we discussed:

> Make some perf tests

Overall, this PR seems to add some overhead, but adding #4318 on top of this PR, the performance becomes better. I think, that's what we are interested in the end. This is aligned with my expectations.


pgbench   -f /tmp/f.sql -c 32 -j 8  -T 300 -P 1 |  
-- | --
Branch | TPS
master | tps = 70.671640
#4317 | tps = 64.681610
#4317   + #4318 | tps = 76.204711



The query is a simple, `SELECT count(*) FROM table`, which triggers `HasIncompleteConnectionEstablishment ()` as I added some logging to check.

> What happens if connection timeout happens while HasIncompleteConnectionEstablishment

The timeout cannot happen as to get into `HasIncompleteConnectionEstablishment ()`, the execution should have guaranteed to finish at least one task. If the execution can finish 1 task, it means that we have at least one successful connection. If we have a successful connection, the timeout cannot.

In general, any other connection failure would also not have any serious affect, such as failing the execution. It could only trigger some `WARNING`s, which seems OK to me.

Reply:
What's the test set up? Is this single-node or multi-node? I guess main thing is not to regress on the latter.

Reply:
Sorry, I should have shared the specs of the test earlier. 

It was 1 coordinator +2 worker Azure VM cluster, where the coordinator is not added to the metadata. PG 12.3 is used, without debug mode (e.g., no assertions).

>         "coordinatorVmSize": {
>             "value": "Standard_D16s_v3"
>         },
>         "workerVmSize": {
>             "value": "Standard_E16s_v3"
 >        },

The table had few million rows, doing `SELECT count(*) FROM table`. I think it was around ~50-80 msecs query when executed in isolation. It was definitely triggering `HasIncompleteConnectionEstablishment ()` as I added some logs to verify.

Reply:
I did one more benchmark with the exact same hardware specs.


Also `pgbench -f /tmp/f.sql -c XXX -j 16  -T 300 -P 1` and `max_connections:   1000`.  I get some mixed results, which I re-run twice to make sure that it is consistent, and they are consistent. 

  | v9.5.0 | #4317
-- | -- | --
c:1 | tps = 26.446103 | tps = 26.077251
c:32 | tps = 160.283905 | tps = 152.059204
c:128 | tps = 209.165868 | tps = 229.989242

So, for a single client, no difference. For medium load, v9.5.0 is %5 better. For larger loads, this PR is %7-8 faster.

I need to dive a bit more on why this is happening. Wanted to share as an heads up

Reply:
Looking into the perf regressions a little more, I think this might have significant implications.  For example, when the table is empty:


  | v9.5.0 | #4317
-- | -- | --
c:1 | tps = 61.464242 | tps = 44.586448